File: backend/app/schemas/activity.py
from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.models.enums import ActivityStatus, ActivityType, Platform


class ActivityRead(BaseModel):
    id: UUID
    account_id: UUID
    platform: Platform
    activity_type: ActivityType
    status: ActivityStatus
    target_url: str | None = None
    title: str | None = None
    metadata_: dict[str, Any] | None = Field(default=None, serialization_alias="metadata")
    started_at: datetime | None = None
    finished_at: datetime | None = None
    duration_ms: int | None = None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True, populate_by_name=True)

File: frontend/src/types/activity.ts
import type { Platform } from "./account";

export type ActivityStatus = "PENDING" | "RUNNING" | "SUCCESS" | "FAILED" | "CANCELLED";

export type ActivityType =
  | "LOGIN"
  | "OPEN_BROWSER"
  | "OPEN_HOME"
  | "BROWSE"
  | "UPVOTE"
  | "DOWNVOTE"
  | "COMMENT"
  | "POST"
  | "JOIN_SUBREDDIT"
  | "LEAVE_SUBREDDIT"
  | "SYNC_PROFILE"
  | "VALIDATE_SESSION"
  | "REFRESH_SESSION"
  | "DELETE_SESSION";

export interface Activity {
  id: string;
  account_id: string;
  platform: Platform;
  activity_type: ActivityType;
  status: ActivityStatus;
  target_url: string | null;
  title: string | null;
  metadata: Record<string, unknown> | null;
  started_at: string | null;
  finished_at: string | null;
  duration_ms: number | null;
  created_at: string;
  updated_at: string;
}

export interface ActivityFilters {
  accountId?: string;
  limit?: number;
  offset?: number;
  activity_type?: ActivityType;
  status?: ActivityStatus;
}
